ord_schema/interface/query.py

```diff
@@ -25,6 +27,155 @@
 # pylint: disable=too-many-locals
 
 
+class Predicate:
+    """Structure and code generation for ORD query predicates."""
+
+    class MatchMode(enum.Enum):
+        """Interpretations for SMILES and SMARTS strings."""
+        EXACT = 1
+        SIMILAR = 2
+        SUBSTRUCTURE = 3
+        SMARTS = 4
+
+        @classmethod
+        def from_name(cls, name):
+            """Take a matching criterion from a URL param."""
+            return Predicate.MatchMode[name.upper()]
+
+    def __init__(self):
+        # List of pairs, (SMILES/SMARTS, MatchMode).
+        self.inputs = []
+        # A pair, (SMILES/SMARTS, MatchMode).
+        self.output = None
+        # A "reaction" SMILES (RDKit fingerprint string).
+        self.reaction_smiles = None
+        # PK strings, "ord-<hex>".
+        self.reaction_ids = []
+        # Threshold for similarity matching.
+        self.tanimoto_threshold = 0.4
+        # Whether to consider stereochemistry.
+        self.do_chiral_ss = False
+
+    def add_input(self, pattern, match_mode):
+        self.inputs.append((pattern, match_mode))
+
+    def set_output(self, pattern, match_mode):
+        self.output = (pattern, match_mode)
+
+    def add_reaction_id(self, reaction_id):
+        self.reaction_ids.append(reaction_id)
+
+    def set_reaction_smiles(self, reaction_smiles):
+        self.reaction_smiles = reaction_smiles
+
+    def set_tanimoto_threshold(self, threshold):
+        self.tanimoto_threshold = threshold
+
+    def use_stereochemistry(self):
+        self.do_chiral_ss = True
+
+    def _inputs_predicate(self):
+        exprs = []
+        for pattern, mode in self.inputs:
+            if mode == Predicate.MatchMode.EXACT:
+                expr = f"inputs.smiles = '{pattern}'"
+            elif mode == Predicate.MatchMode.SIMILAR:
+                expr = f"rdki.mfp2%morganbv_fp('{pattern}')"
+            elif mode == Predicate.MatchMode.SUBSTRUCTURE:
+                expr = f"rdki.m@>'{pattern}'"
+            elif mode == Predicate.MatchMode.SMARTS:
+                expr = f"rdki.m@>'{pattern}'::qmol"
+            exprs.append(expr)
+        return ' AND '.join(exprs) or 'TRUE'
+
+    def _output_predicate(self):
```

Comment:
I think input and output queries are similar enough that they shouldn't duplicate code (they only differ in the table name).